src/vcard/mod_vcard_api.erl

```diff
@@ -0,0 +1,258 @@
+%% @doc Provide an interface for frontends (like graphql or ctl) to manage vcard.
+-module(mod_vcard_api).
+
+-include("mongoose.hrl").
+-include("jlib.hrl").
+-include("mod_vcard.hrl").
+
+-type vcard_map() :: #{binary() => vcard_subelement_binary() | vcard_subelement_map()}.
+-type vcard_subelement_binary() :: binary() | [{ok, binary()}].
+-type vcard_subelement_map() :: #{binary() => binary() | [{ok, binary()}]}.
+
+-export([set_vcard/2,
+         get_vcard/1]).
+
+-spec set_vcard(jid:jid(), vcard_map()) ->
+     {ok, vcard_map()} | {not_found, string()} | {internal, string()} | {vcard_not_found, string()}.
+set_vcard(#jid{luser = LUser, lserver = LServer} = UserJID, Vcard) ->
+    case mongoose_domain_api:get_domain_host_type(LServer) of
+        {ok, HostType} ->
+            case set_vcard(HostType, UserJID, Vcard) of
+                ok ->
+                    get_vcard_from_db(HostType, LUser, LServer);
+                _ ->
+                    {internal, "Internal server error"}
+            end;
+        _ ->
+            {not_found, "User does not exist"}
+    end.
+
+-spec get_vcard(jid:jid()) ->
+     {ok, vcard_map()} | {not_found, string()} | {internal, string()} | {vcard_not_found, string()}.
+get_vcard(#jid{luser = LUser, lserver = LServer}) ->
+    case mongoose_domain_api:get_domain_host_type(LServer) of
+        {ok, HostType} ->
+            get_vcard_from_db(HostType, LUser, LServer);
+        _ ->
+            {not_found, "User does not exist"}
+    end.
+
+set_vcard(HostType, UserJID, Vcard) ->
+    mod_vcard:unsafe_set_vcard(HostType, UserJID, transform_from_map(Vcard)).
+
+get_vcard_from_db(HostType, LUser, LServer) ->
+    ItemNotFoundError = mongoose_xmpp_errors:item_not_found(),
+    case mod_vcard_backend:get_vcard(HostType, LUser, LServer) of
+        {ok, Result} ->
+            [#xmlel{children = VcardData}] = Result,
+            {ok, to_map_format(VcardData)};
+        {error, ItemNotFoundError} ->
+            {vcard_not_found, "Vcard for user not found"};
+        _ ->
+            {internal, "Internal server error"}
```

Comment:
Ok, I was wrong where. I didn't notice the line:
https://github.com/esl/MongooseIM/blob/11cd084b7d27e47bbd9fc2c6ad1f862521972bd8/src/vcard/mod_vcard_api.erl#L44

It can stay as it is ;)